Human vs Human

codeidoscope opened the pull request:
This PR implements a Human VS Human Tic Tac Toe game.

Please note that there is no handling of errors or edge cases.

Commit:
Set up project and create board

Commit:
Merge branch 'master' of https://github.com/codeidoscope/tic-tac-toe-clojure

Commit:
Delete unnecessary cljs objects

Commit:
Rename and refactor board-creating function

Commit:
Add and test display-board function

Commit:
Update main namespace in project.clj

Commit:
Refactor display-board and create numbered-board

Commit:
Create set-position

Commit:
Create get-user-position and -swap-player

Commit:
Create join-sections and three-aligned

Commit:
Create start-game, take-turn loop and game-over?

Commit:
Add instructions to README

Commit:
Refactor display-board, create build-board, fix tests

Commit:
Fix next-player-turn

Commit:
Implement recursion in next-player-turn

Commit:
Fix swapping of player

Commit:
Implement recursion break with game-over?

Commit:
Update cell numbering to match indices

Commit:
Fix defs and tests

Commit:
Remove .nrepl-port file

laurentqro commented on spec/clj/tic_tac_toe_clojure/core_spec.clj:
```diff
@@ -0,0 +1,98 @@
+        ["X" "_" "_" "_" "_" "_" "_" "_" "_"]
+        (set-position (create-board) 0 "X")))
+
+  (it "sets a X mark O mark at the correct position"
```
`sets an X mark and an O mark` :)

fabientownsend replied:
I would even say `sets an X amrk and then an O mark` 

Also, when devleoping O is kind of sticky as it could be seen as a 0 (zerro). Other example, I and l (i and L)

codeidoscope replied:
@fabientownsend - Is there a way around that? Do I just need to mention that this is the letter O? Github is kind enough to cross the 0 so they don't look the same, but I'm not sure how else to avoid that.

fabientownsend replied:
You could call the marks `cross` & `nought`, maybe something like that:
```
(def nought "O")
(def cross "X")
```

RabeaGleissner commented on src/clj/tic_tac_toe_clojure/core.clj:
```diff
@@ -0,0 +1,74 @@
+(defn three-aligned? [board]
+  (some true? (for [section (join-sections board)] (symbols-equal? section))))
+
+(defn three-mapped? [board]
```
I'm not really sure about this method name. What does it mean? Mapped to what?

codeidoscope replied:
@RabeaGleissner - this is a redundant function, it does the same as `three-aligned` and I'd originally written it in an attempt to understand how mapped worked. I will remove it and use `three-aligned` instead!

fabientownsend commented on README.md:
```diff
@@ -1 +1,18 @@
+
+A Tic Tac Toe game written in Clojure.
+
+- [] Human VS Human
```
I think that Markdown expect a space between the brackets `[ ]`

fabientownsend commented on README.md:
```diff
@@ -1 +1,18 @@
+
+## Running the programme
+
+It's currently broken, but whatever tiny part of it works will run by entering the command `lein run`.
```
👍

fabientownsend commented on spec/clj/tic_tac_toe_clojure/core_spec.clj:
```diff
@@ -0,0 +1,98 @@
+(describe "A board"
+  (it "has 9 cells"
+      (should=
+        ["_" "_" "_" "_" "_" "_" "_" "_" "_"]
```
Is there a way to indent that in order to make it look like more a board?
```
[
 "_" "_" "_" 
 "_" "_" "_" 
 "_" "_" "_"
]
```
?

fabientownsend commented on src/clj/tic_tac_toe_clojure/core.clj:
```diff
@@ -0,0 +1,74 @@
+    (println end-game)
+    (next-player-turn (set-position board (get-user-position) user-symbol) (swap-player user-symbol))))
+
+(defn take-turn [board user-symbol]
```
Is that dead code? If so, I think you can delete it from your repository

You can either keep it locally or in the worst case, still can find it in your git history

fabientownsend commented on src/clj/tic_tac_toe_clojure/core.clj:
```diff
@@ -0,0 +1,74 @@
+(def numbered-board
+  ["0" "1" "2" "3" "4" "5" "6" "7" "8"])
+
+(defn build-board [board]
```
it feels like that build board is quite vague - I can understand from the cade that it generate something for the UI though - do you think you could find a more expensive name?

fabientownsend reviewed (commented):
Good job 👍

It may be good that you read about the Single Responsibility Principle, at the moment everything is in the same file/place

andrewMacmurray commented on spec/clj/tic_tac_toe_clojure/core_spec.clj:
```diff
@@ -0,0 +1,98 @@
+  (it "gets the diagonals from a board"
+      (should=
+        [["X" "O" "X"] ["O" "O" "O"]]
+        (get-diagonals (set-position (set-position (set-position (set-position (set-position (create-board) 0 "X") 4 "O") 8 "X") 2 "O") 6 "O"))))
```
Maybe you could use the threading macro to make this a little cleaner? Something like:

```clj
(-> (create_board)
    (set-position 0 "X")
    (set-position 4 "O") ... etc)
```

Or you could clean this up by reducing over a vector of moves, something like:

```clj
(reduce
   (fn [current_board, [position, symbol]] (set-position current_board position symbol)) 
   (create_board)
   [[0 "X"] [4 "O"] [8 "X"]])
```

andrewMacmurray commented on src/clj/tic_tac_toe_clojure/core.clj:
```diff
@@ -0,0 +1,74 @@
+
+(defn get-diagonals [board]
+  [[(nth board 0) (nth board 4) (nth board 8)]
+   [(nth board 2) (nth board 4) (nth board 6)]])
```
Could you use a helper for this like:

```clj
(defn- tiles-at [board moves]
   (map (fn [move] (nth board move)) moves)
```

And then use like:

```clj
(defn get-diagonals [board]
   [(tiles-at [0 4 8])
    (tiles-at [2 4 6])])
```

Or even shorter:

```clj
(defn- tiles-at [board moves]
   (map (partial nth board) moves))

(defn get-diagonals [board]
   (map tiles-at [[0 4 8] [2 4 6]]))
```

Commit:
Update README, remove phantomjs and amend tests and code

codeidoscope commented:
@fabientownsend @RabeaGleissner @andrewMacmurray @matthewglover - Updated.

- I've removed redundant code: phantomjs config, `take-turn` and `three-mapped` functions
- I renamed `build-board` to `format-board`
- I've updated `symbols-equal?` and `get-diagonals`
- I've removed the use of `set-position` in tests where it wasn't necessary and replaced it with a simple vector formatted to look like a board so it's clearer what the test is doing and reduces cognitive overload (this also removes the need to create a helper or use a threading macro)
